src/deploy/util.js

```diff
@@ -1,19 +1,20 @@
-const path = require('path')
+const { basename, sep } = require('path')
 
 const pWaitFor = require('p-wait-for')
 
 // Default filter when scanning for files
-const defaultFilter = (filename) => {
-  if (filename == null) return false
-  const n = path.basename(filename)
-  switch (true) {
-    case n === 'node_modules':
-    case n.startsWith('.') && n !== '.well-known':
-    case n.match(/(\/__MACOSX|\/\.)/):
-      return false
-    default:
-      return true
+const defaultFilter = (filePath) => {
+  if (filePath == null) {
+    return false
   }
+
+  const filename = basename(filePath)
+  return (
+    filename !== 'node_modules' &&
+    !(filename.startsWith('.') && filename !== '.well-known') &&
+    !filename.includes('/__MACOSX') &&
```

Comment:
The logic in lines `15` and `16` is kept the same as the original.
However, the original had a bug it seemed, since `filename` cannot start with `/`. 

I am not 100% what the intent is here. My guess is:
  - do not deploy `node_modules`
  - do not deploy any dot files, except `.well-known`
  - do not deploy `__MACOSX`

I am not quite sure how the recursion applies. For example, if `node_modules` is excluded, does this function gets called with `node_modules/example-module/`? Depending on the answer to this question, the test would be different.

We might also consider using [`junk`](https://github.com/sindresorhus/junk).

@erezrokah What are your thoughts on all this?

It'd be nice to improve this, but I don't want to break anything either.

Reply:
See relevant CLI PR https://github.com/netlify/cli/pull/1272 and the logic the `js-client` is trying to mimic:
https://github.com/netlify/open-api/blob/e72e2eb2d7eedf02bccc5916fc0330022f7f823b/go/porcelain/deploy.go#L692

> * do not deploy any dot files, except `.well-known`

An important distinction is when the publish directory is hidden https://github.com/netlify/cli/issues/1227#issuecomment-696961664 

> I am not quite sure how the recursion applies

I think once a directory is excluded, so does its nested directories:
https://github.com/okdistribute/folder-walker/blob/9c3b07320e152def5c3cbc26c772ebebc6d5c124/index.js#L44 

I think you're correct with the intent and the changes you made makes sense.

Not sure if we want to improve the fix from the CLI as a breaking change. The CLI has some special handling when deploying the root directory as described in the issue.

Reply:
This seems like a breaking change that we might want to hold on for the moment, especially before the holidays.
The initial intent was to improve the coding style, so let's merge this PR with no change of behavior, even if the code is somewhat nonsensical (since it replicates the original code), then follow-up with another PR for the behavior part.